### PR TITLE
Improve build script

### PR DESCRIPTION
**Pull request by granders**
Ended up refactoring the original script quite a bit, and trying to clean it up for improved readability/maintainability.

Changes include
* `-h | --help` for help message
* broke up logic into more self-contained functions
* consolidate git checkout/pull/revert logic
* only wipe out working tree and checkout a specified branch after a fresh download, or when `--update` flag is set

### CODE REVIEW

**Comment by granders** [comment]
@hachikuji @ewencp Mind taking a look?

File: build.sh [hunk 1]
```diff
@@ -32,151 +48,215 @@ while [ $# -gt 0 ]; do
...
 # don't make it into a release branch.
-REVERT_GITHASHES="d04a918a239b6d894f71975b6e1e6dcc390424a3 2b382d6f92ed1d6f414529ce8d41089110ed560a"
-CHERRYPICK_GITHASHES="8db55618d5d5d5de97feab2bf8da4dc45387a76a 04827e6e999c1b5e89e7dc8b1573ad263e66cd56 c6b8de4e6806d8f9f4af57e15f2a7f4170265c42 1408c670ea576a430287182e9aaa26793917129a 969d0cb0ae316ba0dfdb34ed096bfd56fe86ad92 e32df4722ca476ab373cca45a824f712cf4943ef"
-
-if [ ! -d $projects_dir/kafka ]; then
-    echo "Cloning Kafka"
-    git clone http://git-wip-us.apache.org/repos/asf/kafka.git $projects_dir/kafka
-fi
+readonly REVERT_GITHASHES="d04a918a239b6d894f71975b6e1e6dcc390424a3 2b382d6f92ed1d6f414529ce8d41089110ed560a"
+readonly CHERRYPICK_GITHASHES="8db55618d5d5d5de97feab2bf8da4dc45387a76a 04827e6e999c1b5e89e7dc8b1573ad263e66cd56 c6b8de4e6806d8f9f4af57e15f2a7f4170265c42 1408c670ea576a430287182e9aaa26793917129a 969d0cb0ae316ba0dfdb34ed096bfd56fe86ad92 e32df4722ca476ab373cca45a824f712cf4943ef"
```
Comment (hachikuji) [inline build.sh]: Would it make sense to split these commits and some of the versions into a separate file which perhaps gets sourced in this script?

File: build.sh [hunk 2]
```diff
@@ -32,151 +48,215 @@ while [ $# -gt 0 ]; do
...
+    fi
+}
+
+function build_kafka {
+    cd $PROJECTS_DIR/kafka
 
-popd
+    # Default JAVA_HOME for EC2
+    if [ -z "$JAVA_HOME" ]; then
+        export JAVA_HOME=/usr/lib/jvm/java-7-openjdk-amd64
```
Comment (ewencp) [inline build.sh]: This seems like it's potentially brittle. I don't think anything else depends on this having been evaluated, but refactoring this into a function even though the export will continue beyond the scope of the function hides important ordering from the top-level ordering of the scripts operations. An alternative would be to not export it here and just include it for the commands that need it (which I assume are the gradle lines).

**Comment by ewencp** [comment]
LGTM. This does leave a sort of weird mix between globals and function args (e.g. plenty of places use globals, but `apply_kafka_githashes` passes values in from globals despite only being used once). I wouldn't worry too much about further refactoring for now -- it's nice to clean this up and there was a real issue that needed to be fixed with the way we were applying those cherry-picked commits and handling the PS patches, but otherwise this works well for now.

File: build.sh (hunk 2 above)
Comment (granders) [inline build.sh]: Makes sense, good catch

File: build.sh (hunk 1 above)
Comment (granders) [inline build.sh]: Point taken... probably the same point would apply to the other globals which are essentially also "configurations", but will leave as-is for now

**Pull request merged**